src/overlays/actors/ovl_Obj_Timeblock/z_obj_timeblock.c

```diff
@@ -27,41 +57,317 @@ const ActorInit Obj_Timeblock_InitVars = {
+}
+
+void ObjTimeblock_ToggleSwitchFlag(GlobalContext* globalCtx, s32 flag) {
+    if (Flags_GetSwitch(globalCtx, flag)) {
+        Flags_UnsetSwitch(globalCtx, flag);
+    } else {
+        Flags_SetSwitch(globalCtx, flag);
+    }
+}
+
+extern s32 D_06000B30;
+
+void ObjTimeblock_Init(Actor* thisx, GlobalContext* globalCtx) {
+    ObjTimeblock* this = THIS;
+    s32 pad;
+    UNK_TYPE dynaPoly_p = 0;
+
+    DynaPolyInfo_SetActorMove(&this->dyna, 0);
+    this->dyna.actor.shape.rot.z = 0;
+    this->dyna.actor.posRot.rot.z = this->dyna.actor.shape.rot.z;
+
+    DynaPolyInfo_Alloc(&D_06000B30, &dynaPoly_p);
+
+    this->dyna.dynaPolyId =
+        DynaPolyInfo_RegisterActor(globalCtx, &globalCtx->colCtx.dyna, &this->dyna.actor, dynaPoly_p);
+
+    Actor_ProcessInitChain(&this->dyna.actor, sInitChain);
+    Actor_SetScale(&this->dyna.actor, sSizeOptions[(this->dyna.actor.params >> 8) & 1].scale);
+
+    if ((this->dyna.actor.params >> 6) & 1) {
+        this->save = OBJ_TIMEBLOCK_DO_NOT_SAVE;
+    } else {
+        this->save = ((this->dyna.actor.params & 0x3F) < 0x38) ? OBJ_TIMEBLOCK_SAVE_AGE_INVERTED : OBJ_TIMEBLOCK_SAVE;
+    }
+
+    this->songObserverFunc = &ObjTimeblock_WaitForOcarina;
+
+    Actor_SetHeight(&this->dyna.actor, sSizeOptions[(this->dyna.actor.params >> 8) & 1].height);
+
+    this->unk_174 = BOOL(Flags_GetSwitch(globalCtx, this->dyna.actor.params & 0x3F));
+    this->unk_175 = BOOL((this->dyna.actor.params >> 15) & 1);
+    this->isVisible = ObjTimeblock_CalculateIsVisible(this);
+
+    if (!((this->dyna.actor.params >> 10) & 1)) {
+        ObjTimeblock_StartNormalActionFunc(this);
+    } else if (this->isVisible) {
+        ObjTimeblock_StartAltBehaviorVisibleActionFunc(this);
+    } else {
+        ObjTimeblock_StartAltBehaviourNotVisibleActionFunc(this);
+    }
+
+    osSyncPrintf(BLOCK_OF_TIME_JA " (<arg> %04xH <type> save:%d color:%d range:%d move:%d)\n",
+                 (u16)this->dyna.actor.params, this->save, this->dyna.actor.initPosRot.rot.z & 7,
+                 (this->dyna.actor.params >> 11) & 7, (this->dyna.actor.params >> 10) & 1);
+}
+
+void ObjTimeblock_Destroy(Actor* thisx, GlobalContext* globalCtx) {
+    DynaPolyInfo_Free(globalCtx, &globalCtx->colCtx.dyna, THIS->dyna.dynaPolyId);
+}
+
+extern s32 func_80043590(DynaPolyActor* actor);
+
+u8 ObjTimeblock_PlayerIsInRange(ObjTimeblock* this, GlobalContext* globalCtx) {
+    if (this->isVisible && func_80043590(&this->dyna)) {
+        return false;
+    }
+
+    if (this->dyna.actor.xzDistFromLink <= sRanges[(this->dyna.actor.params >> 11) & 7]) {
+        Vec3f distance;
+        f32 blockSize;
+
+        // Will not return true if player would be inside the block?
+        func_8002DBD0(&this->dyna.actor, &distance, &(PLAYER)->actor.posRot.pos);
+        blockSize = this->dyna.actor.scale.x * 50.0f + 6.0f;
+        if (blockSize < fabsf(distance.x) || blockSize < fabsf(distance.z)) {
```

Comment:
```suggestion
        func_8002DBD0(&this->dyna.actor, &distance, &(PLAYER)->actor.posRot.pos);
        blockSize = this->dyna.actor.scale.x * 50.0f + 6.0f;
        // Return true if player's xz position is not inside the block
        if (blockSize < fabsf(distance.x) || blockSize < fabsf(distance.z)) {
```

func_8002DBD0 returns the player's distance from the block, relative to the block's orientation

Reply:
Thanks for the info. Can I use that to document the function?

```patch
--- a/src/code/z_actor.c
+++ b/src/code/z_actor.c
@@ -934,6 +934,7 @@ f32 func_8002DBB0(Actor* actor, Vec3f* arg1) {
     return Math_Vec3f_DistXZ(&actor->posRot.pos, arg1);
 }
 
+// Calculates arg2's distance from actor, relative to the actor's orientation
 void func_8002DBD0(Actor* actor, Vec3f* result, Vec3f* arg2) {
     f32 cosRot2Y;
     f32 sinRot2Y;
```